File: _includes/code/howto/search.basics.py
# ================================
# ===== INSTANTIATION-COMMON =====
# ================================

# ===== Instantiation
import weaviate
import weaviate.classes as wvc
import os

client = weaviate.connect_to_wcs(
    cluster_url=os.getenv("WCS_DEMO_URL"),
    auth_credentials=weaviate.auth.AuthApiKey(os.getenv("WCS_DEMO_RO_KEY")),
    headers={
        "X-OpenAI-Api-Key": os.getenv("OPENAI_APIKEY"),
    }
)

try:

    # ==============================
    # ===== BASIC GET EXAMPLES =====
    # ==============================

    # BasicGetPython
    jeopardy = client.collections.get("JeopardyQuestion")
    # highlight-start
    response = jeopardy.query.fetch_objects()
    # highlight-end

    for o in response.objects:
        print(o.properties)
    # END BasicGetPython

    # TESTS IN THIS FILE NOT CHECKED OR EXPECTED TO RUN YET

    # Test results
    assert response.objects[0].collection == "JeopardyQuestion"
    assert "question" in response.objects[0].properties.keys()
    # End test

    # ====================================
    # ===== BASIC GET LIMIT EXAMPLES =====
    # ====================================

    # GetWithLimitPython
    jeopardy = client.collections.get("JeopardyQuestion")
    response = jeopardy.query.fetch_objects(
        # highlight-start
        limit=1
        # highlight-end
    )

    for o in response.objects:
        print(o.properties)
    # END GetWithLimitPython

    # Test results
    assert response.objects[0].collection == "JeopardyQuestion"
    assert "question" in response.objects[0].properties.keys()
    assert len(response.objects) == 1
    # End test


    # ==========================================
    # ===== GET LIMIT WITH OFFSET EXAMPLES =====
    # ==========================================

    # GetWithLimitOffsetPython
    jeopardy = client.collections.get("JeopardyQuestion")
    response = jeopardy.query.fetch_objects(
        # highlight-start
        limit=1,
        offset=1
        # highlight-end
    )

    for o in response.objects:
        print(o.properties)
    # END GetWithLimitOffsetPython

    # Test results
    assert response.objects[0].collection == "JeopardyQuestion"
    assert "question" in response.objects[0].properties.keys()
    assert len(response.objects) == 1
    # End test


    # ==========================================
    # ===== GET OBJECT PROPERTIES EXAMPLES =====
    # ==========================================

    # GetPropertiesPython
    jeopardy = client.collections.get("JeopardyQuestion")
    response = jeopardy.query.fetch_objects(
        # highlight-start
        limit=1,
        return_properties=["question", "answer", "points"]
        # highlight-end
    )

    for o in response.objects:
        print(o.properties)
    # END GetPropertiesPython

    # Test results
    assert response.objects[0].collection == "JeopardyQuestion"
    for prop_name in ["question", "answer", "points"]:
        assert prop_name in response.objects[0].properties.keys()
    # End test


    # ======================================
    # ===== GET OBJECT VECTOR EXAMPLES =====
    # ======================================

    # GetObjectVectorPython
    # highlight-start
    import weaviate.classes as wvc
    # highlight-end

    jeopardy = client.collections.get("JeopardyQuestion")
    response = jeopardy.query.fetch_objects(
        # highlight-start
        include_vector=True,
        # highlight-end
        limit=1
    )

    print(response.objects[0].vector["default"])
    # END GetObjectVectorPython

    # Test results
    assert response.objects[0].collection == "JeopardyQuestion"
    assert type(response.objects[0].vector["default"]) == list
    assert len(response.objects[0].vector["default"]) >= 100
    # End test


    # ==================================
    # ===== GET OBJECT ID EXAMPLES =====
    # ==================================

    # GetObjectIdPython
    # highlight-start
    import weaviate.classes as wvc
    # highlight-end

    jeopardy = client.collections.get("JeopardyQuestion")
    response = jeopardy.query.fetch_objects(
        # Object IDs are included by default with the `v4` client! :)
        limit=1
    )

    for o in response.objects:
        print(o.uuid)
    # END GetObjectIdPython


    # Test results
    from weaviate.collections.queries.base import _WeaviateUUIDInt

    assert response.objects[0].collection == "JeopardyQuestion"
    assert type(response.objects[0].uuid) == _WeaviateUUIDInt
    # End test


    # ==============================
    # ===== GET WITH CROSS-REF EXAMPLES =====
    # ==============================

    # GetWithCrossRefsPython
    # highlight-start
    import weaviate.classes as wvc
    # highlight-end

    jeopardy = client.collections.get("JeopardyQuestion")
    response = jeopardy.query.fetch_objects(
        # highlight-start
        return_references=[
            wvc.query.QueryReference(
                link_on="hasCategory",
                return_properties=["title"]
            ),
        ],
        # highlight-end
        limit=2
    )

    for o in response.objects:
        print(o.properties["question"])
        # print referenced objects
        for ref_obj in o.references["hasCategory"].objects:
            print(ref_obj.properties)
    # END GetWithCrossRefsPython

    # Test results
    assert response.objects[0].collection == "JeopardyQuestion"
    assert len(response.objects[0].references["hasCategory"].objects) > 0
    # END Test


    # ====================================
    # ===== GET WITH METADATA EXAMPLE =====
    # ====================================

    # GetWithMetadataPython
    # highlight-start
    import weaviate.classes as wvc
    # highlight-end

    jeopardy = client.collections.get("JeopardyQuestion")
    response = jeopardy.query.fetch_objects(
        limit=1,
        # highlight-start
        return_metadata=wvc.query.MetadataQuery(creation_time=True)
        # highlight-end
    )

    for o in response.objects:
        print(o.properties)  # View the returned properties
        print(o.metadata.creation_time)  # View the returned creation time
    # END GetWithMetadataPython

    # Test results
    assert response.objects[0].collection == "JeopardyQuestion"
    assert response.objects[0].metadata.creation_time is not None
    # END Test results


    # =========================
    # ===== MULTI-TENANCY =====
    # =========================

    # MultiTenancy
    import weaviate.classes as wvc

    # Connect to the collection
    mt_collection = client.collections.get("WineReviewMT")

    # Get the specific tenant's version of the collection
    # highlight-start
    collection_tenant_a = mt_collection.with_tenant("tenantA")
    # highlight-end

    # Query tenantA's version
    response = collection_tenant_a.query.fetch_objects(
        return_properties=["review_body", "title"],
        limit=1,
    )

    print(response.objects[0].properties)
    # END MultiTenancy


    # Test results
    assert len(response.objects) > 0
    assert response.objects[0].collection == "WineReviewMT"
    # End test


finally:
    client.close()

File: _includes/code/howto/search.basics.ts
import assert from 'assert';

// ===== Instantiation, not shown in snippet
import weaviate, { WeaviateClient } from 'weaviate-client';

const client: WeaviateClient = await weaviate.connectToWeaviateCloud(
  process.env.WCD_URL,
 {
   authCredentials: new weaviate.ApiKey(process.env.WCD_API_KEY),
   headers: {
     'X-OpenAI-Api-Key': process.env.OPENAI_APIKEY,  // Replace with your inference API key
   }
 } 
)

// BasicGetJS // GetWithLimitJS // GetWithLimitOffsetJS // GetPropertiesJS // GetObjectVectorJS // GetObjectIdJS // GetWithCrossRefsJS // GetWithMetadataJS 
const myCollection = client.collections.get('JeopardyQuestion');
// END BasicGetJS // END GetWithLimitJS // END GetWithLimitOffsetJS // END GetPropertiesJS // END GetObjectVectorJS // END GetObjectIdJS // END GetWithCrossRefsJS // END GetWithMetadataJS


// ==============================
// ===== BASIC GET EXAMPLES =====
// ==============================
{
// BasicGetJS

// highlight-start
const result = await myCollection.query.fetchObjects()
// highlight-end
console.log(JSON.stringify(result, null, 2));

// END BasicGetJS
}

// Test
// assert('JeopardyQuestion' in result.data.Get);
// let questionKeys = new Set(Object.keys(result.objects[0].properties));
// assert.deepEqual(questionKeys, new Set(['question']));
// End test


// ===================================
// ===== GET WITH LIMIT EXAMPLES =====
// ===================================
{
// GetWithLimitJS

const result = await myCollection.query.fetchObjects(
  // highlight-start
  { limit: 1 }
  // highlight-end
)
     
console.log(JSON.stringify(result, null, 2));
// END GetWithLimitJS

// Test
// assert('JeopardyQuestion' in result.data.Get);
// assert.equal(result.objects.length, 1);
// questionKeys = new Set(Object.keys(result.objects[0].properties));
// assert.deepEqual(questionKeys, new Set(['question']));
// End test
}

// ===================================
// ===== GET WITH LIMIT AND OFFSET EXAMPLES =====
// ===================================
{
// GetWithLimitOffsetJS

const result = await myCollection.query.fetchObjects({
  // highlight-start
  limit: 1,
  offset: 1,
  // highlight-end
})

console.log(JSON.stringify(result, null, 2));
// END GetWithLimitOffsetJS

// Test
// assert.deepEqual(result.data.Get, {
//   JeopardyQuestion: [
//     {
//       question: 'Pythons are oviparous, meaning they do this',
//     },
//   ],
// });
// End test
}

// ===================================
// ===== GET PROPERTIES EXAMPLES =====
// ===================================
{
// GetPropertiesJS

const result = await myCollection.query.fetchObjects({
  limit: 1,
  // highlight-start
  returnProperties: ['question', 'answer', 'points'],
  // highlight-end
})

console.log(JSON.stringify(result, null, 2));
// END GetPropertiesJS

// Test
// assert('JeopardyQuestion' in result.data.Get);
// assert.equal(result.objects.length, 1);
// questionKeys = new Set(Object.keys(result.objects[0].properties));
// assert.deepEqual(result.data.Get, {
//   JeopardyQuestion: [
//     {
//       answer: 'Jonah',
//       points: 100,
//       question: 'This prophet passed the time he spent inside a fish offering up prayers',
//     },
//   ],
// });
// End test
}

// ======================================
// ===== GET OBJECT VECTOR EXAMPLES =====
// ======================================
{
// GetObjectVectorJS

const result = await myCollection.query.fetchObjects({
  // highlight-start
  includeVector: true,
  // highlight-end
  limit: 1,

})

console.log(JSON.stringify(result, null, 2));
// END GetObjectVectorJS

// Test
// assert('JeopardyQuestion' in result.data.Get);
// assert.equal(result.objects.length, 1);
// let additionalKeys = new Set(Object.keys(result.objects[0].properties._additional));
// assert.deepEqual(additionalKeys, new Set(['vector']));
// End test
}

// ==================================
// ===== GET OBJECT ID EXAMPLES =====
// ==================================
{
// GetObjectIdJS

const result = await myCollection.query.fetchObjects({
  // Object IDs are included by default with the `v3` client! :)
  limit: 1,
})

for (let object of result.objects) {
   console.log(JSON.stringify(object.uuid, null, 2));
}
// END GetObjectIdJS

// Test
// assert('JeopardyQuestion' in result.data.Get);
// assert.equal(result.objects.length, 1);
// additionalKeys = new Set(Object.keys(result.objects[0].properties._additional));
// assert.deepEqual(additionalKeys, new Set(['id']));
// End test
}

// =======================================
// ===== GET WITH CROSS-REF EXAMPLES =====
// =======================================

{
// GetWithCrossRefsJS

const result = await myCollection.query.fetchObjects({
  limit: 2,
  // highlight-start
  returnReferences: [{
    linkOn: 'hasCategory',
    returnProperties: ['title'],
  }]
  // highlight-end
})

result.objects.forEach(item => 
  console.log(JSON.stringify(item.references, null, 2))
);
// END GetWithCrossRefsJS

// Test
// assert('JeopardyQuestion' in result.data.Get);
// assert.equal(result.objects.length, 2);
// const questionValues = new Set(result.data.Get.JeopardyQuestion.map(q => q.question));
// assert.deepEqual(questionValues, new Set([
//   'This prophet passed the time he spent inside a fish offering up prayers',
//   'Pythons are oviparous, meaning they do this',
// ]));
// End test
}

// ===================================
// ===== GET WITH METADATA EXAMPLES =====
// ===================================
{
// GetWithMetadataJS

const result = await myCollection.query.fetchObjects({
  limit: 2,
  // highlight-start
  returnMetadata: ['creationTime', ]
  // highlight-end
})

for (let object of result.objects) {
  console.log(JSON.stringify(object.properties, null, 2));
  console.log(JSON.stringify(object.metadata?.creationTime, null, 2));
}
// END GetWithMetadataJS

// Test
// assert('JeopardyQuestion' in result.data.Get);
// assert.equal(result.objects.length, 1);
// questionKeys = new Set(Object.keys(result.objects[0].properties));
// assert.deepEqual(questionKeys, new Set(['question']));
// assert('creationTimeUnix' in result.objects[0].metadata.creationTime);
// End test
}
// =========================
// ===== MULTI-TENANCY =====
// =========================

// <!-- NEEDS TESTS -->

// MultiTenancy
const myMTCollection = client.collections.get('WineReviewMT');
// highlight-start
const collectionTenantA = myMTCollection.withTenant('tenantA');
// highlight-end

const multiTenantResult = await collectionTenantA.query.fetchObjects({
  limit: 1,
  returnProperties: ['review_body','title']
})

console.log(JSON.stringify(multiTenantResult.objects[0].properties, null, 2));
// END MultiTenancy

// Test results
true
// End test
